### PR TITLE
Fix planning.

### LINKED ISSUE #175: Can't update Computed fields
### Module version

```
main@84afb9afc9ce3e8e4777ca9024bd641ff646c363
```

### Relevant provider source code

```go
tfsdk.Schema{
                 Attributes: map[string]tfsdk.Attribute{
                         "last_updated": {
                                 Type:     types.StringType,
                                 Computed: true,
                         },
                  },
}
```

```go
func (r resource) Update(_ context.Context, _ tfsdk.UpdateResourceRequest, resp tfsdk.UpdateResourceResponse) {
  err := resp.State.SetAttribute(tftypes.NewAttributePath().WithAttributeName("last_updated"), time.Now().String())
  if err != nil {
    panic(err)
  }
}
```

### Expected Behavior

On update, `last_updated` should reflect the current time.

### Actual Behavior

Teraform threw an error about inconsistent results after apply:

```
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to hashicups_order.edu, provider "provider[\"registry.terraform.io/hashicorp/hashicups-pf\"]" produced an
│ unexpected new value: .last_updated: was cty.StringVal("Tuesday, 21-Sep-21 12:34:00 PDT"), but now cty.StringVal("Tuesday, 21-Sep-21
│ 17:11:04 PDT").
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

### Steps to Reproduce

1. Apply a resource
2. Change any value, making it update
3. Apply the resource again

### Explanation

The root cause of this is that our [`markComputedNilsAsUnknown`](https://github.com/hashicorp/terraform-plugin-framework/blob/84afb9afc9ce3e8e4777ca9024bd641ff646c363/tfsdk/serve.go#L587-L605) transform function is just... wrong.

The way it behaves today is that any computed attribute in the proposed new state that is null gets set to unknown. Which lets the provider choose a value for it at apply time, as Computed promises. But! A wrinkle. The proposed new state [will use the state value, not null, for any computed attribute with a non-null value in state](https://github.com/hashicorp/terraform/blob/main/docs/resource-instance-change-lifecycle.md):

> **Proposed New State**: Terraform Core merges the non-null values from the configuration with any computed attribute results in the prior state to produce a combined object that includes both, to avoid each provider having to re-implement that merging logic. Will be null when planning a delete operation.

This means once these computed attributes get a value in state, we no longer mark them as unknown. We even [have a test that confirms that behavior](https://github.com/hashicorp/terraform-plugin-framework/blob/84afb9afc9ce3e8e4777ca9024bd641ff646c363/tfsdk/serve_test.go#L1701-L1735).

However, this is non-obvious behavior and we should _instead_ mark the field as unknown automatically if it's `null` _in the config_ and computed, so that the provider always has the option of changing it whenever Terraform will allow it. We should do this _before_ plan modifiers are run, so providers have the opportunity to override this behavior.

### PR DESCRIPTION
**Pull request by paddycarver**
This fixes #175. Essentially, our behavior did exactly what we thought
it did, but didn't combine with Terraform in the ways we thought it
would. We need to set all computed attributes that are null in the
_config_ to unknown, not all computed attributes that are null in the
_plan_ to unknown, because once a computed attribute has a value in
state, it will not have a null plan again unless the provider modifies
the plan to null explicitly. This means once a computed attribute gets a
value, under the existing code, it can't be updated by the provider.

As part of this change, I moved when the null->unknown transform
happens. It used to happen at the very end, which gave provider
developers no opportunity to override the auto-unknown values with
values the provider may know at plan time, such as a provider-level
default or the combination of other fields, etc. By doing this first,
the provider has the ability to understand that the value is unknown and
replace it with a known value.

I also fixed a bug that would return an error when trying to use
Schema.AttributeAtPath with a path that pointed to an element inside a
list, map, or set nested attribute.

I also fixed plan modification to still run resource-level plan
modification even when the plan is null (i.e., the resource is being
deleted) so that diagnostics can be returned. This is useful, for
example, when a resource _can't_ be deleted on the API, but the provider
is going to just remove it from Terraform's state. This allows provider
developers to surface a warning at plan time to practitioners about this
limitation, letting them know _before_ they apply the change that that
is what's going to happen.

Finally, some tests had nonsensical inputs to them that Terraform could
never produce--a null config that still had planned values. I updated
those tests to have more realistic values by removing the plan and
removing the RequiresReplace set in the response, as a deleted resource
needing to be replaced doesn't super make sense.

### CODE REVIEW

File: tfsdk/serve_test.go [hunk 1]
```diff
@@ -2045,50 +2040,16 @@ func TestServerPlanResourceChange(t *testing.T) {
...
-						"size_gb": tftypes.NewValue(tftypes.Number, 10),
-						"boot":    tftypes.NewValue(tftypes.Bool, false),
-					}),
-				}),
-			}),
+			proposedNewState: tftypes.NewValue(testServeResourceTypeTwoType, nil),
+			config:           tftypes.NewValue(testServeResourceTypeTwoType, nil),
+			resource:         "test_two",
+			resourceType:     testServeResourceTypeTwoType,
+			// when the config is null, that means the resource has
```
Comment (kmoe) [inline tfsdk/serve_test.go:2047]: The purpose of this test case wasn't to model null config behaviour but to ensure diags and RequiresReplace were populated properly when both are set in modifyPlanFunc. Prefer this test case to be fixed by populating the config, and we can deal with the null config case in a separate test.

**Review by bflad — COMMENTED** [review]
As discussed out of band, this type of change raises some questions around practitioner and provider developer experiences. I would personally love if we could spend some time working through those scenarios, if possible.

---

My understanding of the current behavior in the new framework is that it is similar to Terraform Plugin SDK version 2, where provider developers must explicitly opt-in to setting new or unknown values via plan modifications for a Computed-only attribute if it is to change from the current state value.

If correctly done by the provider developer, this will show up correctly in `terraform plan` output, such as `OLD => NEW` or `OLD => (known after apply)`.

If not, they (provider developers and practitioners alike) will receive the `Provider produced inconsistent result after apply` error diagnostic from Terraform CLI. (Caveat: This error is demoted to warning logs with SDKv2 when the attribute is not referenced in another location, which is not also updated.)

Since the new framework has all the information that Terraform CLI uses to determine this error condition, it is possible that the framework could raise its own error diagnostic with its own guidance.

---

My understanding of the proposed behavior with this change is that the framework will by default mark all Computed-only attributes in the plan as unknown. Provider developers would need to explicitly opt-out of the behavior by setting the expected value (even if it is the current state value) via plan modifications.

If done correctly by provider developers, unchanged attributes will show no difference in `terraform plan` output.

If not, the `terraform plan` output will always show `OLD => (known after apply)` for every `output` or other value references.

The `Provider produced inconsistent result after apply` error diagnostic is much harder or not possible to induce in this scenario.

---

In either case, it might be worth writing up some canonical documentation on the manner since it is likely confusing either which way, both for (future) maintainers of this project and provider developers. 👍

File: tfsdk/serve_test.go [hunk 2]
```diff
@@ -751,6 +751,9 @@ func TestServerValidateProviderConfig(t *testing.T) {
 				return
 			}
 			if diff := cmp.Diff(got.Diagnostics, tc.expectedDiags); diff != "" {
+				for _, d := range got.Diagnostics {
+					t.Log(d.Detail)
+				}
```
[the remark this answers is not in the document]
Comment (bflad) [reply tfsdk/serve_test.go:756]: I've added these helpful debugging bits to https://github.com/hashicorp/terraform-plugin-framework/issues/135#issuecomment-928288295 so we can decide how we want to handle them going forward. Just to reduce the diff of this pull request, I'm going to extract them out and we can do them consistently across the board at a later time.

**Review by bflad — APPROVED** [review]
Great teamwork, everyone! Verified this looks as expected with hashicups-pf. 🚀

**Review by kmoe — APPROVED** [review]

**Pull request merged**